r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -76,13 +131,69 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(res)
     }
 
-    /// Constraints for reducing the result of a multiplication mod p, to get an original representation.
+    /// Reducing a NonNativeMulResultGadget back to a non-native field gadget
+    /// in normal form. Assumes that 
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2.
+    /// ``
+    /// where 
+    /// ``
+    ///     bits_per_limb = NonNativeFieldParams::bits_per_limb, 
+    ///     surfeit' =  log(num_limbs + num_adds + 1), 
+    ///     num_limbs = NonNativeFieldParams::num_limbs,
+    /// ``
+    /// and `num_adds` is as in the NonNativeMulResultGadget.
+    // Costs
+    // ``
+    //     C =  len(p) + surfeit' 
+    //          + len(p) + num_limbs^2
+    //          +  (num_groups - 1) * (3 + bits_per_limb + surfeit') + 1
+    // ``
+    // constraints, with `num_groups = Ceil(len(p) / S)` and
```

Comment:
It should be num_groups =  ceil(bit_len(prod) / (S * bits_per_limb)). Note that max_bit_len(prod) = 2*len(p) + len(num_adds(prod) + 1).